core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,477 @@
+                    processSubchannelState(subchannel, stateInfo);
+                  }
+                });
+              }
+            }
+          }
+
+          // remove old subchannels that were not in new address list
+          for (SocketAddress oldAddr : oldAddrs) {
+            if (!newAddrs.contains(oldAddr)) {
+              subchannels.get(oldAddr).shutdown();
+              subchannels.remove(oldAddr);
+            }
+          }
+
+          if (addressIndex.seekTo(previousAddress)) {
+            // start connection attempt at first address
+            if (currentState == CONNECTING || currentState == TRANSIENT_FAILURE) {
+              addressIndex.reset();
+              requestConnection();
+            // TODO: verify desired behavior (ideally, we should restart at first address)
+            // start connection attempt at first address when requested
+            } else if (currentState == IDLE) {
+              addressIndex.reset();
+              SubchannelPicker picker = new RequestConnectionPicker(subchannels.get(addressIndex.getCurrentAddress()));
+              updateBalancingState(IDLE, picker);
+            }
+          } else {
+            // start connection attempt at first address
+            if (currentState == CONNECTING || currentState == TRANSIENT_FAILURE) {
+              requestConnection();
+            // start connection attempt at first address when requested
+            } else if (currentState == IDLE || currentState == READY) {
+              SubchannelPicker picker = new RequestConnectionPicker(subchannels.get(addressIndex.getCurrentAddress()));
+              updateBalancingState(IDLE, picker);
+            }
+          }
+        }
+      });
+    }
+
+    @Override
+    public void handleNameResolutionError(Status error) {
+      for (Subchannel subchannel : subchannels.values()) {
+        subchannel.shutdown();
+        subchannel = null;
+      }
+      subchannels.clear();
+      // NB(lukaszx0) Whether we should propagate the error unconditionally is arguable. It's fine
+      // for time being.
+      updateBalancingState(TRANSIENT_FAILURE, new Picker(PickResult.withError(error)));
+    }
+
+    void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
+        ConnectivityState newState = stateInfo.getState();
+        // If we have shutdown channels or updated addresses, we may still receive state updates
+        // from subchannels. To prevent pickers from returning these obselete subchannels, this logic
+        // is included to check if the current list of active subchannels includes the current one.
+        if (!subchannels.containsValue(subchannel)) {
+          return;
+        }
+        if (newState == SHUTDOWN) {
+          return;
+        }
+        if (newState == IDLE) {
+          helper.refreshNameResolution(); // TODO: check cases for refresh name resolution
```

Comment:
likely we do refreshnameResolution when
1. the previous ready channel that we chose become IDLE
2. we finished iteration and all subchannels had one TF

Reply:
Fixed test cases to reflect this.